app/src/main/java/io/mochahub/powermeter/workoutsession/workoutsessiondialog/NewWorkoutDialog.kt

```diff
@@ -0,0 +1,340 @@
+                myCalendar.get(Calendar.DAY_OF_MONTH)).show()
+        }
+    }
+
+    // //////////////////////////////////////////////////////////////
+    // Interface methods for workout controller
+    // //////////////////////////////////////////////////////////////
+    // Need to use a loop method because it is more reliable then indexing
+    // If we use index's there is a bug when deleting sets/ workouts
+    // that causes out of bounds exceptions
+    override fun onAddSetClicked(workout: Workout) {
+        for (i in 0 until viewModel.workouts.size) {
+            if (viewModel.workouts[i].id == workout.id) {
+                viewModel.workouts[i] = viewModel.workouts[i].addSet(0, WorkoutSet(weight = 0.0, reps = 0))
+                workoutController.setData(viewModel.workouts)
+            }
+        }
+    }
+
+    override fun onRepFocusChanged(workout: Workout, workoutSet: WorkoutSet, value: Int) {
+        for (i in 0 until viewModel.workouts.size) {
+            if (viewModel.workouts[i].id == workout.id) {
+                for (k in 0 until viewModel.workouts[i].sets.size) {
+                    if (viewModel.workouts[i].sets[k].id == workoutSet.id) {
+                        viewModel.workouts[i].sets[k] = viewModel.workouts[i].sets[k].setReps(value)
+                        workoutController.setData(viewModel.workouts)
+                    }
+                }
+            }
+        }
+    }
+
+    override fun onWeightFocusChanged(workout: Workout, workoutSet: WorkoutSet, value: Double) {
+        for (i in 0 until viewModel.workouts.size) {
+            if (viewModel.workouts[i].id == workout.id) {
+                for (k in 0 until viewModel.workouts[i].sets.size) {
+                    if (viewModel.workouts[i].sets[k].id == workoutSet.id) {
+                        viewModel.workouts[i].sets[k] = viewModel.workouts[i].sets[k].setWeight(value)
+                        workoutController.setData(viewModel.workouts)
+                    }
+                }
+            }
+        }
+    }
+
+    override fun onExerciseSelected(workout: Workout, exercise: String) {
+
+        val foundExercise = exercises.find { it.name == exercise }
+        if (foundExercise == null) {
+            Log.e(this.javaClass.canonicalName, "Exercise not found")
+            return
+        }
+
+        for (i in 0 until viewModel.workouts.size) {
+            if (viewModel.workouts[i].id == workout.id) {
+                viewModel.workouts[i] = viewModel.workouts[i]
+                    .updateExercise(Exercise(foundExercise.name, foundExercise.personalRecord, foundExercise.muscleGroup))
+                workoutController.setData(viewModel.workouts)
+            }
+        }
+    }
+}
+
+private class WorkoutRowSetSwipeCallBack(
+    private val resources: Resources,
+    private val onLeftSwipe: (Int, Int) -> Unit
+) : EpoxyTouchHelper.SwipeCallbacks<WorkoutRowSetModel>() {
```

Comment:
Can we extract this class to its own file?

Reply:
Extracted here https://github.com/power-meter/android-client/pull/32/commits/a348f7c56ca351c997994a15963b67edccfc7cff